java/code/src/com/redhat/rhn/common/hibernate/HibernateFactory.java

```diff
@@ -794,36 +820,39 @@ protected static void executeCallableMode(String name, String mode, Map params)
     @SuppressWarnings("unchecked")
     protected static <T, ID> List<T> findByIds(List<ID> ids, String queryName,
             String idsParameterName, Map<String, Object> parameters) {
-        Session session = HibernateFactory.getSession();
-        org.hibernate.query.Query<T> query = session.getNamedQuery(queryName);
+        Query<T> query = HibernateFactory.getSession().getNamedQuery(queryName);
+
         parameters.entrySet().stream().forEach(entry -> {
             query.setParameter(entry.getKey(), entry.getValue());
         });
-        List<T> results = new LinkedList<T>();
 
-        if (ids.size() == 0) {
-            return results;
-        }
+        return splitAndExecuteQuery(ids, idsParameterName, query, query::getResultList,
+                new ArrayList<T>(), ListUtils::union);
+    }
 
-        if (ids.size() < LIST_BATCH_MAX_SIZE) {
-            query.setParameterList(idsParameterName, ids);
-            return query.getResultList();
-        }
+    /**
+     * Splits a list of elements in batches of LIST_BATCH_MAX_SIZE and execute a query for each batch.
+     * Results from each query are reduced via `accumulator` using the provided `identity`.
+     * @param <T> the return type
+     * @param <E> the type of the elements in the list parameter
+     * @param <R> the type of the returned objects by the query
+     * @param list the list of parameters to search for
+     * @param parameterName the name of the parameter to match the parameters in the list
+     * @param query the query to be executed
+     * @param queryFunction the function to be call on the query
+     * @param identity the identity for the accumulator function
+     * @param accumulator the operation for the result accumulator
+     * @return an accumulated result of executing the query
+     */
+    private static <E, T, R> T splitAndExecuteQuery(List<E> list, String parameterName,
+            Query<R> query, Supplier<T> queryFunction, T identity, BinaryOperator<T> accumulator) {
+        int size = list.size();
```

Comment:
NitPick: If size if zero, could we return the "identity"? We don't need to execute anything else in this scenario.

Reply:
if size == zero, then the "batches" list is empty so nothing will get executed. I would prefer to keep this method as short and simple as possible.